onnxruntime/contrib_ops/cpu/bert/qembed_layer_norm.cc

```diff
@@ -0,0 +1,242 @@
+                    "Beta zero point must be a scalar or 1D tensor of size 1");
+
+  bool has_segment_embedding = segment_embedding != nullptr;
+
+  const int32_t* input_ids_data = input_ids->template Data<int32_t>();
+  const int32_t* segment_ids_data =
+      has_segment_embedding ? segment_ids->template Data<int32_t>() : nullptr;
+
+  const auto& input_dims = input_ids->Shape().GetDims();
+  int batch_size = static_cast<int>(input_dims[0]);
+  int sequence_length = static_cast<int>(input_dims[1]);
+  int64_t hidden_size = word_embedding->Shape()[1];
+
+  int word_embedding_length = static_cast<int>(word_embedding->Shape()[0]);
+  int position_embedding_length = static_cast<int>(position_embedding->Shape()[0]);
+  int segment_embedding_length =
+      has_segment_embedding ? static_cast<int>(segment_embedding->Shape()[0]) : 0;
+
+  // Grab quantization values:
+  float word_embedding_scale = *(word_embedding_scale_tensor->template Data<float>());
+  uint8_t word_embedding_zero_point =
+      *(word_embedding_zero_point_tensor->template Data<uint8_t>());
+
+  float position_embedding_scale = *(position_embedding_scale_tensor->template Data<float>());
+  uint8_t position_embedding_zero_point =
+      *(position_embedding_zero_point_tensor->template Data<uint8_t>());
+
+  float segment_embedding_scale =
+      has_segment_embedding ? *(segment_embedding_scale_tensor->template Data<float>()) : 0.0f;
+  uint8_t segment_embedding_zero_point =
+      has_segment_embedding ? *(segment_embedding_zero_point_tensor->template Data<uint8_t>()) : 0;
+
+  float layer_norm_weights_scale = *(gamma_scale_tensor->template Data<float>());
+  uint8_t layer_norm_weights_zero_point = *(gamma_zero_point_tensor->template Data<uint8_t>());
+
+  float layer_norm_bias_scale = *(beta_scale_tensor->template Data<float>());
+  uint8_t layer_norm_bias_zero_point = *(beta_zero_point_tensor->template Data<uint8_t>());
+
+  // Request outputs:
+  TensorShape output_shape({batch_size, sequence_length, hidden_size});
+  Tensor* output = context->Output(0, output_shape);
+
+  TensorShape mask_index_shape({batch_size});
+  Tensor* mask_index = context->Output(1, mask_index_shape);
+
+  // Grab pointers to buffers each Tensor represents:
+  const uint8_t* word_embedding_data = word_embedding->template Data<uint8_t>();
+  const uint8_t* position_embedding_data = position_embedding->template Data<uint8_t>();
+  const uint8_t* segment_embedding_data =
+      has_segment_embedding ? segment_embedding->template Data<uint8_t>() : nullptr;
+  const uint8_t* gamma_data = gamma->template Data<uint8_t>();
+  const uint8_t* beta_data = beta->template Data<uint8_t>();
+
+  T* output_data = output->template MutableData<T>();
+
+  // Perform the Op:
+  {
+    std::atomic_bool failed{false};
+
+    int n = batch_size * sequence_length;
+    concurrency::ThreadPool::TryBatchParallelFor(
+        context->GetOperatorThreadPool(), n, [=, &failed](ptrdiff_t index) {
```

Comment:
I did play around with this a bit - there is duplicated logic sadly from `EmbedLayerNorm<T>::Compute()`. I think we'd have to break up the logic more. The start offsets are calculated from each batch `word_col_index`, `position_col_index`, and `segment_col_index`.

Happy to tackle this in a fast-follow. Also curious as to what @tlh20 thinks as well!

Reply:
Re @mrry's comment on perf : ThreadPool::TryBatchParallelFor is parameterized on the type of the body (https://github.com/microsoft/onnxruntime/blob/9f1e61be921a5e2e94d89e71f36eca893b457f03/include/onnxruntime/core/platform/threadpool.h#L291-L292).  After template expansion the lambda for the body should then be inlined in a loop over a batch of iterations, as opposed to actually being invoked as a std::function ( https://github.com/microsoft/onnxruntime/blob/9f1e61be921a5e2e94d89e71f36eca893b457f03/include/onnxruntime/core/platform/threadpool.h#L329-L331).  

There are a three potential downsides: (1) Not sure if anyone has explored the code size impact of the template expansion here.  (2) TryBatchParallelFor will split work over the number of threads available.  If the work in each iteration is uneven then we risk load imbalance.  (3) Similarly, if the work per iteration is very small, then we still incur the cost of splitting over all of the threads (when perhaps running all iterations on a single thread may be preferable).

ThreadPool::TryParallelFor aims to address these downsides by having the user provide a cost estimate of the work in each iteration, and then using that to decide how to partition the work. 

I agree further investigation does not need to be a blocker for the PR.  I'll follow up with folk here and @RandySheriffH who is interested in improving the heuristics in the basic ThreadPool::TryParallelFor.

Reply:
Thanks Tim! Somehow I failed to notice the `F` template... 

This might be interesting to look at in a follow-up PR if we see a significant amount of time in these functions. From Tim's list of downsides, I'd actually be most curious about (3), because for the intended deployment of the model that'll use this code we're probably not going to want to use all cores, and single-thread latency would be useful to understand.